src/main/java/edu/greenblitz/tobyDetermined/subsystems/swerve/SwerveChassis.java

```diff
@@ -0,0 +1,459 @@
+			desaturatedStates[i] = new SwerveModuleState(states[i].speedMetersPerSecond / desaturationFactor, states[i].angle);
+		}
+		return desaturatedStates;
+	}
+	
+	public ChassisSpeeds getChassisSpeeds() {
+		return kinematics.toChassisSpeeds(getModuleState(Module.FRONT_LEFT),
+				getModuleState(Module.FRONT_RIGHT),
+				getModuleState(Module.BACK_LEFT),
+				getModuleState(Module.BACK_RIGHT));
+	}
+	
+	public SwerveModulePosition[] getSwerveModulePositions() {
+		return new SwerveModulePosition[]{
+				frontLeft.getCurrentPosition(),
+				frontRight.getCurrentPosition(),
+				backLeft.getCurrentPosition(),
+				backRight.getCurrentPosition()
+		};
+	}
+	
+	public SwerveDriveKinematics getKinematics() {
+		return this.kinematics;
+	}
+	
+	public PigeonGyro getPigeonGyro() {
+		return pigeonGyro;
+	}
+	
+	/**
+	 * moving a single module by module state
+	 */
+	private void setModuleStateForModule(Module module, SwerveModuleState state) {
+		getModule(module).setModuleState(state);
+	}
+	
+	/**
+	 * rotates chassis by percentOutput [-1, 1]
+	 */
+	public void rotateChassisByPower(double percentOutput) {
+		for (int i = 0; i < Module.values().length; i++) {
+			Translation2d translationFromCenter = RobotMap.Swerve.SwerveLocationsInSwerveKinematicsCoordinates[i];
+			getModule(Module.values()[i]).rotateToAngle(Math.atan2(translationFromCenter.getY(), translationFromCenter.getX()) + Math.PI * 0.5);
+			getModule(Module.values()[i]).setLinPowerOnlyForCalibrations(percentOutput);
+		}
+	}
+	
+	/**
+	 * for calibration purposes
+	 */
+	public void rotateModuleByPower(Module module, double power) {
+		getModule(module).setRotPowerOnlyForCalibrations(power);
+	}
+	
+	public void updatePoseEstimationPhotonVision() {
+		poseEstimator.update(getPigeonAngle(), getSwerveModulePositions());
+		Photonvision.getInstance().getUpdatedPoseEstimator().ifPresent((EstimatedRobotPose pose) -> poseEstimator.addVisionMeasurement(pose.estimatedPose.toPose2d(), pose.timestampSeconds));
+	}
+	
+	public void updatePoseEstimationLimeLight() {
+		if(MultiLimelight.getInstance().getFirstAvailableTarget().isPresent() || odometry.getPoseMeters().getTranslation().getDistance(getRobotPose().getTranslation()) < RobotMap.Odometry.MAX_DISTANCE_TO_FILTER_OUT) {
+			if(!robotHasObstacles()) {
+				poseEstimator.update(getGyroAngle(), getSwerveModulePositions());
+			}
+		}
+		if (doVision) {
+			if (MultiLimelight.getInstance().getAllEstimates().size() >= 1) {
+				twoApriltagsPresent = true;
+				for (Optional<Pair<Pose2d, Double>> target : MultiLimelight.getInstance().getAllEstimates()) {
+					target.ifPresent(this::addVisionMeasurement);
+				}
+			}
+		}
+	}
+	private boolean moduleHasObstacles(SwerveModule module) {
+		return module.getMotorOutputCurrent() > RobotMap.Swerve.SdsSwerve.FREE_CURRENT - currentTolerance && module.getMotorOutputCurrent() < currentTolerance + RobotMap.Swerve.SdsSwerve.FREE_CURRENT;
+	}
+	public boolean robotHasObstacles() {
```

Comment:
tal - "hasObstacles" does not make it clear that you mean has-something-that-makes-it-harder-for-it-to-move. maybe say ModulesStalling or smthn

Reply:
maybe RestrictiveMovement / obstructive